File: lib/git_test.go
package lib

import (
	"github.com/alexcesaro/log"
	"github.com/alexcesaro/log/golog"
	"github.com/stretchr/testify/assert"
	"io/ioutil"
	"os"
	"strings"
	"testing"
	"time"
)

var (
	logger  golog.Logger
	git     *Git
	execGit func(...string) error
)

func init() {
	logger = *golog.New(os.Stderr, log.Warning)

	exec := &Exec{
		logger: logger,
		Cmd:    "git",
	}
	execGit = func(args ...string) error {
		err := exec.CommandRun(args...)
		if err == nil {
			return nil
		}
		panic(err)
	}
	git = NewGit(logger)
}

func TestGit_IsInstalled(t *testing.T) {
	setup()
	defer cleanup()

	path := os.Getenv("PATH")
	defer os.Setenv("PATH", path)
	os.Setenv("PATH", ".")

	assert.False(t, git.IsInstalled())
}

func TestGit_IsRepositoryWhenMissing(t *testing.T) {
	setup()
	defer cleanup()

	assert.False(t, git.IsRepo())
}

func TestGit_IsRepositoryFromSubdirectory(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	if err := os.MkdirAll("x/s/a", 0744); err != nil {
		panic(err)
	}
	if err := os.Chdir("x/s/a"); err != nil {
		panic(err)
	}

	assert.True(t, git.IsRepo())
}

func TestGit_IsNotDirtyGitRepository(t *testing.T) {
	setup()
	defer cleanup()

	execGit("init")

	assert.False(t, git.IsDirty())
}

func TestGit_IsDirty(t *testing.T) {
	setup()
	defer cleanup()

	execGit("init")

	_, err := os.Create("test")
	if err != nil {
		assert.Error(t, err)
	}

	assert.True(t, git.IsDirty())
}

func TestGit_BranchExists(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	execGit("checkout", "-b", "test")

	assert.True(t, git.BranchExists("test"))
}

func TestGit_BranchExistsNotOnBranch(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	execGit("checkout", "-b", "test")
	execGit("checkout", "master")

	assert.True(t, git.BranchExists("test"))
}

func TestGit_BranchExistsRmote(t *testing.T) {
	setup()
	defer cleanup()

	// remote repository
	os.Mkdir("remote", 0755)
	os.Chdir("remote")

	execGit("init")
	execGit("config", "user.email", "test@ci.com")
	execGit("config", "user.name", "test")
	os.Create("test")
	execGit("add", "test")
	execGit("commit", "-m", "'test'", "test")
	execGit("checkout", "-b", "test")
	parent, _ := os.Getwd()

	// local
	os.Chdir("..")
	execGit("clone", parent, "local")
	os.Chdir("local")
	git.Fetch()

	assert.True(t, git.BranchExists("test"))
}

func TestGit_BranchCheckoutNew(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	git.BranchCheckoutNew("test")

	assert.True(t, git.BranchExists("test"), "missing branch test")
	assert.Equal(t, git.BranchCurrent(), "test")
}

func TestGit_BranchCheckoutExisting(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()
	execGit("checkout", "-b", "test")
	execGit("checkout", "master")

	git.BranchCheckoutExisting("test")

	assert.True(t, git.BranchExists("test"), "missing branch test")
	assert.Equal(t, git.BranchCurrent(), "test")
}

func TestGit_Commit(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	_, err := os.Create("pom.xml")
	if err != nil {
		panic(err)
	}

	git.Commit("initial update")

	assert.False(t, git.IsDirty())
}

func TestGit_IsInSyncWithMasterSame(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	execGit("checkout", "-b", "test")

	assert.True(t, git.IsInSyncWith("master"))
}

func TestGit_IsInSyncWithMasterAhead(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	execGit("checkout", "-b", "test")
	os.Create("test2")
	execGit("add", "test2")
	execGit("commit", "-m", "'test'", "test2")

	assert.True(t, git.IsInSyncWith("master"))
}

func TestGit_IsNotInSyncWithMaster(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	os.Create("test2")
	execGit("add", "test2")
	execGit("commit", "-m", "'test'", "test2")

	execGit("checkout", "-b", "test")
	execGit("reset", "--hard", "HEAD~1")

	assert.False(t, git.IsInSyncWith("master"))
	assert.False(t, git.IsDirty())
}

func TestGit_MergeMasterIntoBranch(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	os.Create("test2")
	execGit("add", "test2")
	execGit("commit", "-m", "'test'", "test2")

	execGit("checkout", "-b", "test")
	execGit("reset", "--hard", "HEAD~1")

	os.Create("test2")
	ioutil.WriteFile("test2", []byte("test"), 0700)
	execGit("add", "test2")
	execGit("commit", "-m", "'test in branch test2'", "test2")

	assert.False(t, git.IsInSyncWith("master"))
	assert.False(t, git.HasMergeConflict("master"))

	// action
	git.DominantMergeFrom("master", "updates from master")

	assert.True(t, git.IsInSyncWith("master"))
}

func TestGit_MergeMasterIntoBranchWithConflict(t *testing.T) {
	setup()
	defer cleanup()

	createRepoWithSingleCommit()

	// write to test in branch test
	execGit("checkout", "-b", "test")
	ioutil.WriteFile("test", []byte("test"), 0700)
	execGit("add", "test")
	execGit("commit", "-m", "'update test'", "test")

	// write to test in master different
	execGit("checkout", "master")
	ioutil.WriteFile("test", []byte("afafd"), 0700)
	execGit("add", "test")
	execGit("commit", "-m", "'update test'", "test")

	execGit("checkout", "test")

	assert.True(t, git.HasMergeConflict("master"))

	// action
	git.DominantMergeFrom("master", "updates from master")

	time.Sleep(10 * time.Millisecond)

	assert.True(t, git.IsInSyncWith("master"))
	output, _ := ioutil.ReadFile("test")
	n := len(output)
	content := strings.TrimSpace(string(output[:n]))
	assert.Equal(t, "afafd", content)
}

func createRepoWithSingleCommit() {
	execGit("init")

	execGit("config", "user.email", "test@ci.com")
	execGit("config", "user.name", "test")

	os.Create("test")
	execGit("add", "test")
	execGit("commit", "-m", "'test'", "test")
}

File: src/test/java/de/lgohlke/mavenupgrade/GitTest.java
package de.lgohlke.mavenupgrade;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.assertThat;

public class GitTest {
    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();

    private Git git;
    private Path path;
    private Exec gitExec;

    @Before
    public void setUp2() throws Exception {
        path = temporaryFolder.newFolder().toPath();
        git = new Git(path);
        gitExec = new Exec("git", path);
        assertThat(git.isInstalled()).isTrue();
    }

    @Test
    public void shouldDetectMissingRepository() throws Exception {
        assertThat(git.isRepo()).isFalse();
    }

    @Test
    public void shouldDetectRepositoryWhenInSubdirectory() throws Exception {
        createRepoWithSingleCommit();

        Path subdir = path.resolve(Paths.get("x", "s", "a"));
        subdir.toFile().mkdirs();

        Git git = new Git(subdir);
        assertThat(git.isRepo()).isTrue();
    }

    @Test
    public void IsNotDirtyGitRepository() throws Exception {
        execGit("init");

        assertThat(git.isDirty()).isFalse();
    }

    @Test
    public void IsDirtyGitRepository() throws Exception {
        execGit("init");

        path.resolve("test").toFile().createNewFile();

        assertThat(git.isDirty()).isTrue();
    }


    @Test
    public void BranchExists() throws IOException {
        createRepoWithSingleCommit();

        execGit("checkout", "-b", "test");

        assertThat(git.BranchExists("test")).isTrue();
    }

    @Test
    public void BranchExistsNotOnBranch() throws IOException {
        createRepoWithSingleCommit();

        execGit("checkout", "-b", "test");
        execGit("checkout", "master");

        assertThat(git.BranchExists("test")).isTrue();
    }

    private void execGit(String... args) {
        Exec.Result result = gitExec.exec(args);
        if (result.hasError()) {
            throw new IllegalStateException(result.getStderrLines());
        }
    }

    private void createRepoWithSingleCommit() throws IOException {
        execGit("init");

        execGit("config", "user.email", "test@ci.com");
        execGit("config", "user.name", "test");

        path.resolve("test").toFile().createNewFile();

        execGit("add", "test");
        execGit("commit", "-m", "'test'", "test");
    }

    @Test
    public void BranchExistsRemote() throws Exception {
        // remote repository
        Path remote = path.resolve("remote");
        remote.toFile().mkdir();

        gitExec = new Exec("git", remote);
        execGit("init");
        execGit("config", "user.email", "test@ci.com");
        execGit("config", "user.name", "test");
        remote.resolve("test").toFile().createNewFile();
        execGit("add", "test");
        execGit("commit", "-m", "'test'", "test");
        execGit("checkout", "-b", "test");
        execGit("checkout", "master");

        // local
        gitExec = new Exec("git", path);
        execGit("clone", remote.toString(), "local");

        Git gitExec = new Git(path.resolve("local"));
        assertThat(gitExec.BranchExists("test")).isTrue();
    }

    @Test
    public void BranchCheckoutNew() throws Exception {
        createRepoWithSingleCommit();

        git.BranchCheckoutNew("test");

        assertThat(git.BranchExists("test")).as("missing branch test").isTrue();
        assertThat(git.BranchCurrent()).isEqualTo("test");
    }

    @Test
    public void BranchCheckoutExisting() throws Exception {
        createRepoWithSingleCommit();
        execGit("checkout", "-b", "test");
        execGit("checkout", "master");

        git.BranchCheckoutExisting("test");

        assertThat(git.BranchExists("test")).as("missing branch test").isTrue();
        assertThat(git.BranchCurrent()).isEqualTo("test");
    }

    @Test
    public void Commit() throws Exception {
        createRepoWithSingleCommit();

        path.resolve("pom.xml").toFile().createNewFile();

        git.Commit("initial update");

        assertThat(git.isDirty()).isFalse();
    }

    @Test
    public void IsInSyncWithMasterSame() throws Exception {
        createRepoWithSingleCommit();

        execGit("checkout", "-b", "test");

        assertThat(git.IsInSyncWith("master")).isTrue();
    }

    @Test
    public void IsInSyncWithMasterAhead() throws Exception {
        createRepoWithSingleCommit();

        execGit("checkout", "-b", "test");
        path.resolve("test2").toFile().createNewFile();
        execGit("add", "test2");
        execGit("commit", "-m", "'test'", "test2");

        assertThat(git.IsInSyncWith("master")).isTrue();
    }

    @Test
    public void IsNotInSyncWithMaster() throws Exception {
        createRepoWithSingleCommit();

        path.resolve("test2").toFile().createNewFile();
        execGit("add", "test2");
        execGit("commit", "-m", "'test'", "test2");

        execGit("checkout", "-b", "test");
        execGit("reset", "--hard", "HEAD~1");

        assertThat(git.IsInSyncWith("master")).isFalse();
        assertThat(git.isDirty()).isFalse();
    }

    @Test
    public void MergeMasterIntoBranch() throws Exception {
        createRepoWithSingleCommit();

        path.resolve("test2").toFile().createNewFile();
        execGit("add", "test2");
        execGit("commit", "-m", "'test'", "test2");

        execGit("checkout", "-b", "test");
        execGit("reset", "--hard", "HEAD~1");

        Files.write(path.resolve("test2"), "test".getBytes());
        execGit("add", "test2");
        execGit("commit", "-m", "'test in branch test2'","test2");

        assertThat(git.IsInSyncWith("master")).isFalse();
        assertThat(git.HasMergeConflict("master")).isFalse();

        // action
        git.DominantMergeFrom("master", "updates from master");

        assertThat(git.IsInSyncWith("master")).isTrue();
    }

    @Test
    public void MergeMasterIntoBranchWithConflict() throws Exception {
        createRepoWithSingleCommit();

        // write to test in branch test
        execGit("checkout", "-b", "test");

        Files.write(path.resolve("test"), "test".getBytes());
        execGit("add", "test");
        execGit("commit", "-m", "'update test'", "test");

        // write to test in master different
        execGit("checkout", "master");

        Files.write(path.resolve("test"), "afafd".getBytes());
        execGit("add", "test");
        execGit("commit", "-m", "'update test'", "test");

        execGit("checkout", "test");

        assertThat(git.HasMergeConflict("master")).isTrue();

        // action
        git.DominantMergeFrom("master", "updates from master");

        TimeUnit.MILLISECONDS.sleep(10);

        assertThat(git.IsInSyncWith("master")).isTrue();
        assertThat("afafd".getBytes()).isEqualTo(Files.readAllBytes(path.resolve("test")));
    }
}
